Misc/NEWS.d/next/Core and Builtins/2018-09-20-21-36-57.bpo-34589.9zERCD.rst

```diff
@@ -0,0 +1,7 @@
+The locale coercion implementation has been adjusted back to being closer to
+the design documented in PEP 538. This means locale coercion is once again
+solely the responsibility of the application embedding the CPython runtime
```

Comment:
As released in 3.7.0, locale coercion happens at the point where `Py_Initialize` gets called (as that then internally coerces the locale to something else). That's appropriate for UTF-8 mode (since that only affects the CPython runtime in the current process, and the code carefully goes back and recalculates the settings that were already determined based on the locale encoding), but not for locale coercion (since that affects all subsequently executed locale-aware code in the current process, as well as in subprocesses, but can't reliably go back and rewrite all previously determined values in the embedding application).

That's not the way PEP 538 was originally implemented, and deliberately so: instead, the locale coercion happened in *the CLI app*, before the first call to `Py_DecodeLocale`, so nothing would ever get decoded as ASCII in the first place.

While that CLI app code has since moved to a private API inside the shared library (as `_Py_UnixMain`), this PR still gets it back to the same relative location: before the first call to `Py_DecodeLocale`, so its effectively indistinguishable from `LC_CTYPE=C.UTF-8` actually being set externally in the environment.

Reply:
A potential analogy for why the way 3.7 is handling this is broken (and why PEP 538 deliberately defined it as working differently): performing locale coercion in `Py_Main` and `Py_Initialize` is akin to a Python library reconfiguring the logging module as a side effect of doing something else.

Reply:
Amusingly enough, from the comments in the `readline` module implementation:

```
/* GNU readline() mistakenly sets the LC_CTYPE locale.
 * This is evil.  Only the user or the app's main() should do this!
 * We must save and restore the locale around the rl_initialize() call.
 */
```